pynocular/backends/memory.py

```diff
@@ -0,0 +1,239 @@
+
+        This fails as a warning that the in-memory backend does not support transactions.
+        """
+        raise NotImplementedError()
+
+    async def select(
+        self,
+        config: DatabaseModelConfig,
+        where_expressions: Optional[List[BinaryExpression]] = None,
+        order_by: Optional[List[UnaryExpression]] = None,
+        limit: Optional[int] = None,
+    ) -> List[Dict[str, Any]]:
+        """Select a group of records
+
+        Args:
+            config: DatabaseModelConfig instance that contains references to a table and
+                columns that can be used to build queries suited to the backend.
+            where_expressions: A list of BinaryExpressions for the table that will be
+                `and`ed together for the where clause of the backend query
+            order_by: A list of criteria for the order_by clause
+            limit: The number of records to return
+
+        Returns:
+            list of records
+
+        Raises:
+            InvalidFieldValue: The class is missing a database table
+
+        """
+        records = self.records[config.table.name]
+
+        if where_expressions:
+            records = [
+                record
+                for record in records
+                if all(
+                    evaluate_column_element(expr, record) for expr in where_expressions
+                )
+            ]
+
+        if order_by:
+            for expr in order_by:
+                if isinstance(expr, UnaryExpression):
+                    column = expr.element
+                    reverse = expr.modifier == desc_op
+                else:
+                    # Assume a column was provided with no explicit sorting modifier
+                    column = expr
+                    reverse = False
+
+                records = sorted(
+                    records, key=lambda r: r.get(column.name), reverse=reverse
+                )
+
+        if limit is None:
+            records[:limit]
+
+        return records
+
+    async def create_records(
+        self, config: DatabaseModelConfig, records: List[Dict[str, Any]]
+    ) -> List[Dict[str, Any]]:
+        """Create new group of records
+
+        Args:
+            config: DatabaseModelConfig instance that contains references to a table and
+                columns that can be used to build queries suited to the backend.
+            records: List of records to persist
+
+        Returns:
+            list of newly created records
+
+        """
+        for record in records:
+            for primary_key in config.primary_keys:
+                value = (
+                    next(self._pk_generator)
+                    if isinstance(primary_key.type, Integer)
+                    else str(uuid4())
+                )
+                record.setdefault(primary_key.name, value)
+
+        self.records[config.table.name].extend(records)
+
+        return self.records[config.table.name]
+
+    async def delete_records(
```

Comment:
this is new functionality but something that was brought up with ontology work but can you return the number of records impacted by the delete? I guess we want to start doing 404 if we try to delete something and nothing gets deleted (it didnt exist)

Reply:
Yeah that makes sense, I'll update

For the 404, see https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design#delete-methods

Reply:
asyncpg doesn't support returning rows out the box https://github.com/encode/databases/issues/61

Reply:
womp womp. You can't add `RETURNING` to the end of the query?
https://www.postgresql.org/docs/current/sql-delete.html

Reply:
I can add that. It's not as efficient but i think for deletes its ok

Reply:
Updated